@captainsafia - https://dev.azure.com/dnceng/public/_build/results?buildId=1025923&view=ms.vss-test-web.build-test-results-tab&runId=31881838&resultId=125311&paneView=debug

```
C:\h\w\B1000931\w\B2770998\e\sdk338\x64\sdk\6.0.100-preview.3.21155.4\MSBuild.dll -distributedlogger:Microsoft.DotNet.Tools.MSBuild.MSBuildLogger,C:\h\w\B1000931\w\B2770998\e\sdk338\x64\sdk\6.0.100-preview.3.21155.4\dotnet.dll*Microsoft.DotNet.Tools.MSBuild.MSBuildForwardingLogger,C:\h\w\B1000931\w\B2770998\e\sdk338\x64\sdk\6.0.100-preview.3.21155.4\dotnet.dll -maxcpucount -property:Configuration=Release -target:Publish -verbosity:m /bl .\AspNet.n1bwsxr4lfc.Server.csproj
You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
AspNet.n1bwsxr4lfc.Shared -> C:\h\w\B1000931\w\B2770998\e\Templates\BaseFolder\AspNet.n1bwsxr4lfc\Shared\bin\Release\net6.0\AspNet.n1bwsxr4lfc.Shared.dll
CSC : warning CS8785: Generator 'RazorSourceGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'AggregateException' with message 'One or more errors occurred. (Object reference not set to an instance of an object.)' [C:\h\w\B1000931\w\B2770998\e\Templates\BaseFolder\AspNet.n1bwsxr4lfc\Client\AspNet.n1bwsxr4lfc.Client.csproj]
C:\h\w\B1000931\w\B2770998\e\Templates\BaseFolder\AspNet.n1bwsxr4lfc\Client\Program.cs(18,40): error CS0246: The type or namespace name 'App' could not be found (are you missing a using directive or an assembly reference?) [C:\h\w\B1000931\w\B2770998\e\Templates\BaseFolder\AspNet.n1bwsxr4lfc\Client\AspNet.n1bwsxr4lfc.Client.csproj]

Expected: True
```

--- a/src/Components/Components/src/LayoutComponentBase.cs
+++ b/src/Components/Components/src/LayoutComponentBase.cs
@@ -1,6 +1,10 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System.Diagnostics.CodeAnalysis;
+using System.Threading.Tasks;
+using static Microsoft.AspNetCore.Internal.LinkerFlags;
+
 namespace Microsoft.AspNetCore.Components
 {
     /// <summary>
@@ -17,5 +21,12 @@ namespace Microsoft.AspNetCore.Components
         /// </summary>
         [Parameter]
         public RenderFragment? Body { get; set; }
+
+        /// <inheritdoc />
+        // Derived instances of LayoutComponentBase do not appear in any statically analyzable
+        // calls of OpenComponent<T> where T is well-known. Consequently we have to explicitly provide a hint to the trimmer to preserve
+        // properties.
+        [DynamicDependency(Component, typeof(LayoutComponentBase))]
+        public override Task SetParametersAsync(ParameterView parameters) => base.SetParametersAsync(parameters);
     }
 }

--- a/src/Components/Components/src/PublicAPI.Unshipped.txt
+++ b/src/Components/Components/src/PublicAPI.Unshipped.txt
@@ -31,6 +31,7 @@ Microsoft.AspNetCore.Components.CascadingTypeParameterAttribute
 Microsoft.AspNetCore.Components.CascadingTypeParameterAttribute.CascadingTypeParameterAttribute(string! name) -> void
 Microsoft.AspNetCore.Components.CascadingTypeParameterAttribute.Name.get -> string!
 Microsoft.AspNetCore.Components.RenderTree.Renderer.GetEventArgsType(ulong eventHandlerId) -> System.Type!
+override Microsoft.AspNetCore.Components.LayoutComponentBase.SetParametersAsync(Microsoft.AspNetCore.Components.ParameterView parameters) -> System.Threading.Tasks.Task!
 static Microsoft.AspNetCore.Components.ParameterView.FromDictionary(System.Collections.Generic.IDictionary<string!, object?>! parameters) -> Microsoft.AspNetCore.Components.ParameterView
 virtual Microsoft.AspNetCore.Components.RenderTree.Renderer.DispatchEventAsync(ulong eventHandlerId, Microsoft.AspNetCore.Components.RenderTree.EventFieldInfo? fieldInfo, System.EventArgs! eventArgs) -> System.Threading.Tasks.Task!
 *REMOVED*readonly Microsoft.AspNetCore.Components.RenderTree.RenderTreeEdit.RemovedAttributeName -> string

--- a/src/Components/Shared/src/WebEventData.cs
+++ b/src/Components/Shared/src/WebEventData.cs
@@ -1,12 +1,14 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+#nullable enable
+
 using System;
 using System.Diagnostics.CodeAnalysis;
 using System.Text.Json;
 using Microsoft.AspNetCore.Components.RenderTree;
 using static Microsoft.AspNetCore.Internal.LinkerFlags;
-#nullable enable
+
 namespace Microsoft.AspNetCore.Components.Web
 {
     internal class WebEventData
@@ -61,21 +63,13 @@ namespace Microsoft.AspNetCore.Components.Web
         {
             try
             {
-                if (TryGetStandardWebEventArgsType(eventName, out var eventArgsType))
+                if (TryDeserializeStandardWebEventArgs(eventName, eventArgsJson, out var eventArgs))
                 {
-                    // Special case for ChangeEventArgs because its value type can be one of
-                    // several types, and System.Text.Json doesn't pick types dynamically
-                    if (eventArgsType == typeof(ChangeEventArgs))
-                    {
-                        return DeserializeChangeEventArgs(eventArgsJson);
-                    }
-                }
-                else
-                {
-                    // For custom events, the args type is determined from the associated delegate
-                    eventArgsType = renderer.GetEventArgsType(eventHandlerId);
+                    return eventArgs;
                 }
 
+                // For custom events, the args type is determined from the associated delegate
+                var eventArgsType = renderer.GetEventArgsType(eventHandlerId);
                 return (EventArgs)JsonSerializer.Deserialize(eventArgsJson, eventArgsType, JsonSerializerOptionsProvider.Options)!;
             }
             catch (Exception e)
@@ -84,7 +78,7 @@ namespace Microsoft.AspNetCore.Components.Web
             }
         }
 
-        private static bool TryGetStandardWebEventArgsType(string eventName, [MaybeNullWhen(false)] out Type type)
+        private static bool TryDeserializeStandardWebEventArgs(string eventName, string eventArgsJson, [NotNullWhen(true)] out EventArgs? eventArgs)
         {
             // For back-compatibility, we recognize the built-in list of web event names and hard-code
             // rules about the deserialization type for their eventargs. This makes it possible to declare
@@ -97,13 +91,15 @@ namespace Microsoft.AspNetCore.Components.Web
             {
                 case "input":
                 case "change":
-                    type = typeof(ChangeEventArgs);
+                    // Special case for ChangeEventArgs because its value type can be one of
+                    // several types, and System.Text.Json doesn't pick types dynamically
+                    eventArgs = DeserializeChangeEventArgs(eventArgsJson);
                     return true;
 
                 case "copy":
                 case "cut":
                 case "paste":
-                    type = typeof(ClipboardEventArgs);
+                    eventArgs = Deserialize<ClipboardEventArgs>(eventArgsJson);
                     return true;
 
                 case "drag":
@@ -113,20 +109,20 @@ namespace Microsoft.AspNetCore.Components.Web
                 case "dragover":
                 case "dragstart":
                 case "drop":
-                    type = typeof(DragEventArgs);
+                    eventArgs = Deserialize<DragEventArgs>(eventArgsJson);
                     return true;
 
                 case "focus":
                 case "blur":
                 case "focusin":
                 case "focusout":
-                    type = typeof(FocusEventArgs);
+                    eventArgs = Deserialize<FocusEventArgs>(eventArgsJson);
                     return true;
 
                 case "keydown":
                 case "keyup":
                 case "keypress":
-                    type = typeof(KeyboardEventArgs);
+                    eventArgs = Deserialize<KeyboardEventArgs>(eventArgsJson);
                     return true;
 
                 case "contextmenu":
@@ -137,11 +133,11 @@ namespace Microsoft.AspNetCore.Components.Web
                 case "mousedown":
                 case "mouseup":
                 case "dblclick":
-                    type = typeof(MouseEventArgs);
+                    eventArgs = Deserialize<MouseEventArgs>(eventArgsJson);
                     return true;
 
                 case "error":
-                    type = typeof(ErrorEventArgs);
+                    eventArgs = Deserialize<ErrorEventArgs>(eventArgsJson);
                     return true;
 
                 case "loadstart":
@@ -150,7 +146,7 @@ namespace Microsoft.AspNetCore.Components.Web
                 case "load":
                 case "loadend":
                 case "progress":
-                    type = typeof(ProgressEventArgs);
+                    eventArgs = Deserialize<ProgressEventArgs>(eventArgsJson);
                     return true;
 
                 case "touchcancel":
@@ -159,7 +155,7 @@ namespace Microsoft.AspNetCore.Components.Web
                 case "touchenter":
                 case "touchleave":
                 case "touchstart":
-                    type = typeof(TouchEventArgs);
+                    eventArgs = Deserialize<TouchEventArgs>(eventArgsJson);
                     return true;
 
                 case "gotpointercapture":
@@ -172,22 +168,22 @@ namespace Microsoft.AspNetCore.Components.Web
                 case "pointerout":
                 case "pointerover":
                 case "pointerup":
-                    type = typeof(PointerEventArgs);
+                    eventArgs = Deserialize<PointerEventArgs>(eventArgsJson);
                     return true;
 
                 case "wheel":
                 case "mousewheel":
-                    type = typeof(WheelEventArgs);
+                    eventArgs = Deserialize<WheelEventArgs>(eventArgsJson);
                     return true;
 
                 case "toggle":
-                    type = typeof(EventArgs);
+                    eventArgs = Deserialize<EventArgs>(eventArgsJson);
                     return true;
 
                 default:
                     // For custom event types, there are no built-in rules, so the deserialization type is
                     // determined by the parameter declared on the delegate.
-                    type = null;
+                    eventArgs = null;
                     return false;
             }
         }

